### PR TITLE
[Datasets] [Arrow 7+ Support - 4/N] [RFC] Add zero-copy batch API for `ds.map_batches()`.

### LINKED ISSUE #29817: [Datasets] Add zero-copy batch argument for `ds.map_batches()`
For UDFs that don't mutate their input, we should have an option to provide best-effort zero-copy batch views for batch mapping. This can allow us to elide an unnecessary batch copy.

## API

For the best UX, the argument can be tailored to the UDF's batch mutation semantics, such as `allow_mutate_batch: bool = True`. This defaults to `True` to preserve existing behavior, where the batch is guaranteed to be mutable; the docstring for this argument can indicate that disabling the arg will provide best-effort zero-copy view semantics, and can be advantageous for faster execution and lower memory utilization.

### PR DESCRIPTION
**Pull request by clarkzinzow**
By fixing the Arrow serialization bug we were able to remove a lot of the defensive copying throughout the Datasets codebase, enabling us to provide a zero-copy batch optimization path for `ds.map_batches()`. For UDFs that don't mutate a batch, this can elide a batch copy, resulting in lower worker heap memory utilization and better performance.

## API

We have two primary API choices:

1. `allow_mutate_batch: bool`: This API is geared towards the requirements of the user's UDF - if their UDF doesn't mutate the batch, they will set this to `False`, at which point they will get a zero-copy (immutable) batch; if they UDF does mutate the batch, this should be `True` (default) in order to guarantee that their UDF is provided a mutable batch copy.
2. `zero_copy_batch: bool`: This API is geared towards the zero-copy optimization - if the user is looking to optimize their `ds.map_batches()` performance, they will flip this flag to elide a batch copy.

Each has their merits - `allow_mutate_batch` keeps the zero-copy optimization an implementation detail and instead defines the API in terms of UDF requirements that's easy for the user to understand, and may be easier to migrate to a default of `False`; `zero_copy_batch` is more discoverable for devs looking for optimizations, but the immutability aspect of the zero-copy batch then leaks a bit of the Datasets implementation detail.

This PR currently implements option (2), the `zero_copy_batch` API, since the discoverability for users wanting to take advantage of this optimization is better. See the first commit for option (1).

## TODOs

- [x] Update transformation feature guide with tip about this arg.
- [ ] (Follow-up PR) Port preprocessors that don't mutate their input batches to use the `zero_copy_batch=True` arg.
- [ ] (Follow-up PR) Elide batch copy when batch format conversion already creates a copy.
- [ ] (Follow-up PR) Elide batch copy when base block is not in the object store (e.g. upon fusion between read and map_batches tasks).
- [ ] (Follow-up PR) Only create a batch copy when the underlying batch is read-only, without incurring expensive buffer checks.

## Related issue number

Closes https://github.com/ray-project/ray/issues/29817

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -652,15 +692,19 @@ def add_column(
                 ray (e.g., num_gpus=1 to request GPUs for the map tasks).
         """
 
-        def process_batch(batch):
-            batch[col] = fn(batch)
+        def process_batch(batch: "pandas.DataFrame") -> "pandas.DataFrame":
+            batch.loc[:, col] = fn(batch)
```
Comment (clarkzinzow) [inline python/ray/data/dataset.py:696]: Driveby: we should use the `.loc[]` API for mutating DataFrame batches in-place.

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -696,7 +740,11 @@ def drop_columns(
         """
 
         return self.map_batches(
-            lambda batch: batch.drop(columns=cols), compute=compute, **ray_remote_args
+            lambda batch: batch.drop(columns=cols),
+            batch_format="pandas",
+            zero_copy_batch=True,
```
Comment (clarkzinzow) [inline python/ray/data/dataset.py:745]: UDF doesn't mutate in-place.

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -733,6 +781,7 @@ def select_columns(
         """
         return self.map_batches(
             lambda batch: BlockAccessor.for_block(batch).select(columns=cols),
+            zero_copy_batch=True,
```
Comment (clarkzinzow) [inline python/ray/data/dataset.py:784]: UDF doesn't mutate in-place.

File: python/ray/data/dataset.py [hunk 4]
```diff
@@ -461,6 +471,15 @@ def map_batches(
                 ``pandas.DataFrame``, "pyarrow" to select ``pyarrow.Table``, or
                 ``"numpy"`` to select ``numpy.ndarray`` for tensor datasets and
                 ``Dict[str, numpy.ndarray]`` for tabular datasets. Default is "default".
+            zero_copy_batch: Whether ``fn`` should be provided zero-copy, read-only
+                batches. If this is ``True`` and no copy is required for the
+                ``batch_format`` conversion, the batch will be a zero-copy, read-only
```
Comment (jianoaix) [inline python/ray/data/dataset.py:476]: Can we expand a bit about this? like 1) what block types support zero-copy (a Core level feature); 2) what block-to-batch conversions are zero-copy. If we have documentations already, we can just link to it.

**Comment by clarkzinzow** [comment]
@jianoaix comment addressed, @c21 @jianoaix pinging for review!

File: python/ray/data/dataset.py [hunk 5]
```diff
@@ -461,6 +471,17 @@ def map_batches(
...
+                view on data in Ray's object store, which can decrease memory
```
Comment (c21) [inline python/ray/data/dataset.py:477]: Orthogonal to this PR - if `map_batches` and `read` fused together, the block is in-memory and not a view from object store. Can we get rid of copy automatically when block is a view from object store?

**Review by c21 — APPROVED** [review]
LGTM

File: python/ray/data/dataset.py (hunk 5 above)
Comment (clarkzinzow) [reply python/ray/data/dataset.py:477]: Yep I looked into this, among other optimizations!

The semantics are currently very conservative, where we tell the `Batcher` to ensure that batches are always fresh copies (if we're creating batches, i.e. `batch_size is not None`):
* https://github.com/ray-project/ray/blob/8f823cb51b8aee1e1fa8e351c70448d14bf771f5/python/ray/data/dataset.py#L570
* https://github.com/ray-project/ray/blob/8f823cb51b8aee1e1fa8e351c70448d14bf771f5/python/ray/data/_internal/batcher.py#L109-L112

This may create redundant copies if either

1. the base block is already in the worker heap (e.g. due to read and map_batches fusion, or due to the block not supporting zero-copy deserialization), or,
2. the batch conversion will already create a data copy.

The all-encompassing solution would be explicitly checking if the batch is read-only and only then making a copy, but for batch formats like Pandas DataFrames, this is a bit difficult/expensive: we'd need to do an `O(blocks) = O(columns x chunks)` traversal of the DataFrame to get the underlying (ndarray) blocks, and check if each block is writable. This appears to only be possible by using a lot of internal Pandas APIs, which will be super fragile:
```python
if all(block.array_values.to_numpy().flags["WRITEABLE"] for block in batch._data.blocks):
    # No copy needed.
else:
    # Copy needed.
```
This is less fragile if we converted the top-level columns to ndarrays and checked the `"WRITEABLE"` flag for those ndarrays, but this would often trigger block consolidation to create contiguous column ndarrays, resulting in a copy of the DataFrame just to check if we can get away with not copying the DataFrame, which is obviously a non-starter.

Comment (clarkzinzow) [reply python/ray/data/dataset.py:477]: For `zero_copy_batch=False`, I think that the following setup would be the best:
1. For batch formats that support relatively cheap (`O(columns)`) writability checks, such as NumPy and Arrow batches, only create a copy if the batch is not writable.
2. For other batch formats, i.e. Pandas and simple batches, elide copies if we are certain that the base block is already in the worker heap, or if we know that the block format --> batch format conversion will create a full copy.

**Comment by clarkzinzow** [comment]
Failures are unrelated, merging

File: python/ray/data/dataset.py (hunk 5 above)
Comment (c21) [reply python/ray/data/dataset.py:477]: @clarkzinzow - got it, thanks. Maybe in optimizer, we can have a rule to auto-set `zero_copy_batch` if `map_batches` and `read` are fused together.

Comment (clarkzinzow) [reply python/ray/data/dataset.py:477]: @c21 I created an issue to capture this: https://github.com/ray-project/ray/issues/31073
